block-layer/codecs/dag-pb.md

```diff
@@ -36,6 +36,21 @@ message PBNode {
 The objects link names are specified in the 'Name' field of the PBLink object.
 All link names in an object must either be omitted or unique within the object.
 
+This layout can be expressed with [IPLD Schemas](../../schemas/README.md) as:
+
+```ipldsch
+type PBLink struct {
+  hash optional Bytes
+  name optional String
+  tsize optional Int
+}
+
+type PBNode struct {
+  links [PBLink]
+  data Bytes
```

Comment:
`data` *should* be `optional`?  I think PBNode with only lists of links and no data are not uncommon.

Reply:
> I think PBNode with only lists of links and no data are not uncommon.

They aren't valid. You need to signify the "sub-codec" for the links:
- is it a bunch of file fragments? ( type 2, formerly type 0 )
- is it a directory with all links being named ( type 1 )
- is it a single named symlink ( type 4 )
- is it a hamted-directory-shard ( type 5 )

Also due to any intermediate file-stream node ( type 2 ) in a graph being a file on its own right, the size of the stream section and the sync-hints go into `data`

TLDR: not optoinal

Reply:
( above is written in the sense of "there is virtually no use of dag-pb outside of unixfs", if there is such a beast what I wrote above does not apply )

Reply:
For DagPB, data _is_ optional. However, for unixfs, it's _not_ optional.

And yeah, there are uses outside of unixfs (the webui has a dagpb geolocation database, cluster uses it, the pin datastructure uses it, etc.). They're just not commonly transferred/queried on the network (well, except for the geolocation databse).

Reply:
In terms of the _schema_, I wouldn't distinguish between "there is data but its empty" and "there is no data".

Reply:
> I wouldn't distinguish between "there is data but its empty" and "there is no data".

I'm not sure where that leaves us, does an `optional` descriptor help us here at all? I suppose to describe the format, adding an `optional` is telling the reader that there _are_ use-cases where this might not be present. So, I should add an `optional`?

Reply:
Seems we're spec'ing DagPB here, not unixfs(v1), so: on this line we should say `data optional Bytes`, yep.

(And continue to **not** say `nullable` -- saying `nullable` would be claiming there's a distinguishment between "there is data but its empty(/null)" and "there is no data", which per @Stebalien 's statement would be incorrect.)

Reply:
Ah. So you distinguish between "optional" (does not need to be specified in the serialized data) and "nullable" (null doesn't matter)? In that case, everything in dagpb is optional, and nothing is nullable.

Reply:
ok, so `links` is optional too then? or will that just be a zero-length array?

I'm getting a bit mixed up here with what we're trying to represent. We're saying this is the "logical" format but I think this pushes it into the language level where the distinction between `nullable` and `optional` starts to get weird and maybe not so helpful. When we're talking about representation then yes, crisp and clear, but that's not what this is saying. The representation is described by the protobuf spec above this block. I _think_ we should just be aiming for what is most conceptually useful to the reader of this doc trying to understand what data is in a dag-pb block.

Have pushed a new version but I'm dubious, latest is:

```ipldsch
type PBLink struct {
  Hash Bytes
  Name optional String
  Tsize optional Int
}

type PBNode struct {
  Links optional [PBLink]
  Data optional Bytes
}
```

Reply:
> I _think_ we should just be aiming for what is most conceptually useful to the reader of this doc trying to understand what data is in a dag-pb block.

The reason I opened this issue is, that go-ipld-prime transforms from the Data Model into the actual DagPB. Rust IPLD does the same and I'd like to make sure the both are actually doing the same thing/all implementers can just look at the schema, implement it and are good.

Reply:
ok, well, 🤷‍♂, I think this might be good to go but it's a bit hard to tell for sure since concepts are not lining up with crystal clarity